File: cpp/pack.cc
#include <cstdint>
#include <cstdio>
#include <initializer_list>
#include <optional>
#include <sstream>
#include <string>
#include <tuple>
#include <vector>

namespace fmt {

template <typename T, std::enable_if_t<std::is_arithmetic_v<T>, bool> = true>
std::string repr(T value) {
  return std::to_string(value);
}

std::string repr(bool value) { return value ? "true" : "false"; }

// todo: implement tuple
std::string repr(const std::tuple<> &) { return "unit"; }

template <typename T> std::string repr(const std::vector<T> &value) {
  const std::size_t n = value.size();
  std::stringstream s;
  s << "[";
  for (std::size_t i = 0; i < n; ++i) {
    if (i) {
      s << ", ";
    }
    s << repr(value[i]);
  }
  s << "]";
  return s.str();
}

std::string repr(const std::string &value) {
  std::stringstream s;
  s << "\"" << value << "\"";
  return s.str();
}

}; // namespace fmt

namespace pack {
enum type_id : uint8_t {
  type_type = 0x01,
  uint8_type = 0x10,
  uint16_type = 0x11,
  uint32_type = 0x12,
  uint64_type = 0x13,
  int8_type = 0x18,
  int16_type = 0x19,
  int32_type = 0x1a,
  int64_type = 0x1b,
  float_type = 0x20,
  double_type = 0x21,
  bool_type = 0x30,
  list_type = 0x40,
  string_type = 0x41,
};

using Bytes = std::vector<uint8_t>;

class Pack : public Bytes {
public:
  Pack() : Bytes() {}

  template <typename... Ts> Pack(Ts... comps) : Bytes() {
    Bytes comp_data;
    (
        [&]() {
          const Bytes comp_data{comps};
          insert(end(), comp_data.begin(), comp_data.end());
        }(),
        ...);
  }

  Pack(std::initializer_list<uint8_t> init) : Bytes(init) {}

  inline void dump() const {
    uint32_t i = 0;
    for (const auto &byte : *this) {
      if (i) {
        printf("%c", " \n"[!(i % 8)]);
      }
      i++;
      printf("%02x", byte);
    }
    printf("\n");
  }
};

class Unpacker;
template <typename T> inline static Pack pack_bits(const T &value);
template <typename T> inline static std::optional<T> unpack_bits(Unpacker &up);
template <typename T> struct type {
  static const Pack type_info;
  // Directly pack and unpack the bit representation by default
  // This necessarily needs to be specialized for complex types
  inline static Pack pack(const T &value) { return pack_bits(value); }
  inline static std::optional<T> unpack(Unpacker &up) {
    return unpack_bits<T>(up);
  }
};
template <typename T> const Pack type_info = type<T>::type_info;

class Packer {
public:
  template <typename T> inline Packer &pack_value(const T &value) {
    push(type<T>::pack(value));
    return *this;
  }

  template <typename T> inline Packer &pack(const T &value) {
    push(type_info<T>);
    pack_value(value);
    return *this;
  }

  inline void push(const Bytes &data) {
    m_data.insert(m_data.end(), data.begin(), data.end());
  }

  inline const Pack &operator*() { return m_data; }

private:
  Pack m_data{};
};

class Unpacker {
public:
  Unpacker(const Bytes &data) : m_data(data) {};

  template <typename T> inline std::optional<T> unpack_value() {
    return type<T>::unpack(*this);
  }

  template <typename T> inline std::optional<T> unpack() {
    return expect(type_info<T>) ? unpack_value<T>() : std::nullopt;
  }

  inline std::optional<Pack> consume(uint32_t n) {
    if (m_idx + n > m_data.size()) {
      return std::nullopt;
    }
    const Pack data(Bytes(m_data.begin() + m_idx, m_data.begin() + m_idx + n));
    m_idx += n;
    return data;
  }

private:
  inline bool expect(const Bytes &expected) {
    return consume(expected.size()) == expected;
  }

  Pack m_data;
  uint32_t m_idx = 0;
};

template <typename T> inline static Pack pack_bits(const T &value) {
  const uint8_t *value_ptr = reinterpret_cast<const uint8_t *>(&value);
  return Pack(Bytes(value_ptr, value_ptr + sizeof(T)));
}

template <typename T> inline static std::optional<T> unpack_bits(Unpacker &up) {
  const auto data_opt = up.consume(sizeof(T));
  if (!data_opt) {
    return std::nullopt;
  }
  T value{};
  for (uint32_t i = 0; i < sizeof(T); ++i) {
    *(reinterpret_cast<uint8_t *>(&value) + i) = (*data_opt)[i];
  }
  return value;
}

template <> const Pack type<uint8_t>::type_info = {uint8_type};
template <> const Pack type<uint16_t>::type_info = {uint16_type};
template <> const Pack type<uint32_t>::type_info = {uint32_type};
template <> const Pack type<uint64_t>::type_info = {uint64_type};
template <> const Pack type<int8_t>::type_info = {int8_type};
template <> const Pack type<int16_t>::type_info = {int16_type};
template <> const Pack type<int32_t>::type_info = {int32_type};
template <> const Pack type<int64_t>::type_info = {int64_type};
template <> const Pack type<float>::type_info = {float_type};
template <> const Pack type<double>::type_info = {double_type};
template <> const Pack type<bool>::type_info = {bool_type};

template <typename T, std::size_t N> struct type<T[N]> {
  inline static const Pack type_info = Pack(list_type, pack::type_info<T>);
  static Pack pack(const T (&value)[N]) {
    Packer p{};
    p.pack_value<uint32_t>(N);
    for (const auto &elem : value) {
      p.pack_value(elem);
    }
    return *p;
  }
  static std::optional<T[N]> unpack(Unpacker *up) = delete;
};

template <typename T> struct type<std::vector<T>> {
  inline static const Pack type_info = Pack(list_type, pack::type_info<T>);
  static Pack pack(const std::vector<T> &value) {
    Packer p{};
    p.pack_value<uint32_t>(value.size());
    for (const auto &elem : value) {
      p.pack_value(elem);
    }
    return *p;
  }
  static std::optional<std::vector<T>> unpack(Unpacker &up) {
    std::vector<T> value{};
    const auto n_opt = up.unpack_value<uint32_t>();
    if (!n_opt) {
      return std::nullopt;
    }
    for (uint32_t i = 0; i < *n_opt; ++i) {
      const auto elem_opt = up.unpack_value<T>();
      if (!elem_opt) {
        return std::nullopt;
      }
      value.push_back(*elem_opt);
    }
    return value;
  }
};

template <> struct type<std::string> {
  inline static const Pack type_info = {string_type};
  static Pack pack(const std::string &value) {
    Packer p{};
    p.pack_value<uint32_t>(value.size());
    for (const auto &ch : value) {
      p.pack_value<uint8_t>(ch);
    }
    return *p;
  }
  static std::optional<std::string> unpack(Unpacker &up) {
    std::stringstream value{};
    const auto n_opt = up.unpack_value<uint32_t>();
    if (!n_opt) {
      return std::nullopt;
    }
    for (uint32_t i = 0; i < *n_opt; ++i) {
      const auto ch_opt = up.unpack_value<uint8_t>();
      if (!ch_opt) {
        return std::nullopt;
      }
      value << static_cast<char>(*ch_opt);
    }
    return value.str();
  }
};

template <typename... Ts> inline static Pack pack(const Ts &...values) {
  Packer p{};
  (p.pack(values), ...);
  return *p;
}

template <typename T>
inline static std::optional<T> unpack_one(const Bytes &data) {
  return Unpacker(data).unpack<T>();
}

template <typename... Ts, std::size_t... Idx>
inline static std::optional<std::tuple<Ts...>>
process_unpacked(const std::tuple<std::optional<Ts>...> &unpacked,
                 std::index_sequence<Idx...>) {
  return (!std::get<Idx>(unpacked) || ...)
             ? std::nullopt
             : std::make_optional<std::tuple<Ts...>>(
                   {*std::get<Idx>(unpacked)...});
}

template <typename... Ts>
inline static std::optional<std::tuple<Ts...>>
process_unpacked(const std::tuple<std::optional<Ts>...> &unpacked) {
  return process_unpacked(unpacked, std::index_sequence_for<Ts...>{});
}

template <typename... Ts>
inline static std::optional<std::tuple<Ts...>> unpack(const Bytes &data) {
  Unpacker up(data);
  const std::tuple<std::optional<Ts>...> unpacked = {up.unpack<Ts>()...};
  return process_unpacked(unpacked);
}
}; // namespace pack

template <typename T> void test_pack_only(const T &value) {
  printf("input: %s\n", fmt::repr(value).c_str());

  const auto packed = pack::pack(value);
  printf("packed:\n");
  packed.dump();
}

template <typename T> void test(const T &value) {
  printf("input: %s\n", fmt::repr(value).c_str());

  const auto packed = pack::pack(value);
  printf("packed:\n");
  packed.dump();

  const auto unpacked = pack::unpack_one<T>(packed);
  if (!unpacked)
    printf("failed to unpack\n");
  else {
    printf("unpacked: %s\n", fmt::repr(*unpacked).c_str());
  }
}

int main() {
  test<uint32_t>(5);
  printf("\n");
  test<std::vector<int8_t>>({-1, 1, -2, 2, -3, 3, -4, 4});
  printf("\n");
  test(std::string("hello world"));
}

File: py/pack.py
from enum import Enum
import struct

# todo: type annotations?
# todo: unpack<T>() and unpack<(T1, T2)>() syntax?


class type_id(Enum):
    type_type = 0x01
    uint8_type = 0x10
    uint16_type = 0x11
    uint32_type = 0x12
    uint64_type = 0x13
    int8_type = 0x18
    int16_type = 0x19
    int32_type = 0x1A
    int64_type = 0x1B
    float_type = 0x20
    double_type = 0x21
    bool_type = 0x30
    list_type = 0x40
    string_type = 0x41


def resolve_type(value):
    def fail(reason=None):
        if reason:
            raise NotImplementedError(f"{value!r} has unsupported type: {reason}")
        else:
            raise NotImplementedError(f"{value!r} has unsupported type")

    if isinstance(value, int):
        return int32_type

    elif isinstance(value, bool):
        return bool_type

    elif isinstance(value, list):
        if len(value) == 0:
            fail("list with lenght 0")

        elem_type = resolve_type(value[0])
        # todo: implement any(not castable(elem, elem_type) for ...)?
        if any(resolve_type(elem) is not elem_type for elem in value):
            fail("not all elements have the same type")

        return list_type.of(elem_type)

    elif isinstance(value, str):
        return string_type

    fail("not implemented")


class Pack:
    def __init__(self, *comps):
        self.data = bytes()
        for comp in comps:
            if isinstance(comp, Pack):
                self.data += comp.data
            elif isinstance(comp, bytes):
                self.data += comp
            else:
                self.data += bytes((comp,))

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        return self.data[idx]

    def __add__(self, other):
        return Pack(self, other)

    def __eq__(self, other):
        return self.data == Pack(other).data

    def dump(self):
        i = 0
        for byte in self.data:
            if i:
                print("\n "[(i % 8 + 7) // 8], end="")
            i += 1
            print(f"{byte:02x}", end="")
        print()


class Packer:
    def __init__(self):
        self.data = Pack()

    def _push(self, data):
        self.data += data

    def pack_value(self, value, T=None):
        T = T or resolve_type(value)
        assert T is not None
        self._push(T.pack(value))
        return self

    def pack(self, value, T=None):
        T = T or resolve_type(value)
        assert T is not None
        self._push(T.type_info)
        self.pack_value(value, T=T)
        return self


class Unpacker:
    def __init__(self, data):
        self.data = Pack(data)
        self.idx = 0

    def consume(self, n):
        if self.idx + n > len(self.data):
            return None
        data = self.data[self.idx : self.idx + n]
        self.idx += n
        return data

    def _expect(self, expected):
        data = self.consume(len(expected))
        return data == expected

    def unpack_value(self, T):
        return T.unpack(self)

    def unpack(self, T):
        return self.unpack_value(T) if self._expect(T.type_info) else None


# todo: reduce boilerplate
class uint8_type:
    type_info = Pack(type_id.uint8_type.value)

    @staticmethod
    def validate(value):
        assert isinstance(value, int)
        assert 0 <= value < (1 << 8)

    @staticmethod
    def pack(value):
        uint8_type.validate(value)
        return Pack(struct.pack("<B", value))

    @staticmethod
    def unpack(up):
        data = up.consume(1)
        return None if data is None else struct.unpack("<B", data)[0]


class uint16_type:
    type_info = Pack(type_id.uint16_type.value)

    @staticmethod
    def validate(value):
        assert isinstance(value, int)
        assert 0 <= value < (1 << 16)

    @staticmethod
    def pack(value):
        uint16_type.validate(value)
        return Pack(struct.pack("<H", value))

    @staticmethod
    def unpack(up):
        data = up.consume(2)
        return None if data is None else struct.unpack("<H", data)[0]


class uint32_type:
    type_info = Pack(type_id.uint32_type.value)

    @staticmethod
    def validate(value):
        assert isinstance(value, int)
        assert 0 <= value < (1 << 32)

    @staticmethod
    def pack(value):
        uint32_type.validate(value)
        return Pack(struct.pack("<I", value))

    @staticmethod
    def unpack(up):
        data = up.consume(4)
        return None if data is None else struct.unpack("<I", data)[0]


class uint64_type:
    type_info = Pack(type_id.uint64_type.value)

    @staticmethod
    def validate(value):
        assert isinstance(value, int)
        assert 0 <= value < (1 << 64)

    @staticmethod
    def pack(value):
        uint64_type.validate(value)
        return Pack(struct.pack("<Q", value))

    @staticmethod
    def unpack(up):
        data = up.consume(8)
        return None if data is None else struct.unpack("<Q", data)[0]


class int8_type:
    type_info = Pack(type_id.int8_type.value)

    @staticmethod
    def validate(value):
        assert isinstance(value, int)
        assert -(1 << 7) <= value < (1 << 7)

    @staticmethod
    def pack(value):
        int8_type.validate(value)
        return Pack(struct.pack("<b", value))

    @staticmethod
    def unpack(up):
        data = up.consume(1)
        return None if data is None else struct.unpack("<b", data)[0]


class int16_type:
    type_info = Pack(type_id.int16_type.value)

    @staticmethod
    def validate(value):
        assert isinstance(value, int)
        assert -(1 << 15) <= value < (1 << 15)

    @staticmethod
    def pack(value):
        int16_type.validate(value)
        return Pack(struct.pack("<h", value))

    @staticmethod
    def unpack(up):
        data = up.consume(2)
        return None if data is None else struct.unpack("<h", data)[0]


class int32_type:
    type_info = Pack(type_id.int32_type.value)

    @staticmethod
    def validate(value):
        assert isinstance(value, int)
        assert -(1 << 31) <= value < (1 << 31)

    @staticmethod
    def pack(value):
        int32_type.validate(value)
        return Pack(struct.pack("<i", value))

    @staticmethod
    def unpack(up):
        data = up.consume(4)
        return None if data is None else struct.unpack("<i", data)[0]


class int64_type:
    type_info = Pack(type_id.int64_type.value)

    @staticmethod
    def validate(value):
        assert isinstance(value, int)
        assert -(1 << 63) <= value < (1 << 63)

    @staticmethod
    def pack(value):
        int64_type.validate(value)
        return Pack(struct.pack("<q", value))

    @staticmethod
    def unpack(up):
        data = up.consume(8)
        return None if data is None else struct.unpack("<q", data)[0]


class float_type:
    type_info = Pack(type_id.float_type.value)

    # todo: possible to automate checking for castable types?
    @staticmethod
    def validate(value):
        assert isinstance(value, float) or isinstance(value, int)

    @staticmethod
    def pack(value):
        float_type.validate(value)
        return Pack(struct.pack("<f", value))

    @staticmethod
    def unpack(up):
        data = up.consume(4)
        return None if data is None else struct.unpack("<f", data)[0]


class double_type:
    type_info = Pack(type_id.double_type.value)

    # todo: possible to automate checking for castable types?
    @staticmethod
    def validate(value):
        assert isinstance(value, float) or isinstance(value, int)

    @staticmethod
    def pack(value):
        double_type.validate(value)
        return Pack(struct.pack("<d", value))

    @staticmethod
    def unpack(up):
        data = up.consume(8)
        return None if data is None else struct.unpack("<d", data)[0]


class bool_type:
    type_info = Pack(type_id.bool_type.value)

    @staticmethod
    def validate(value):
        assert isinstance(value, bool)

    @staticmethod
    def pack(value):
        bool_type.validate(value)
        return Pack(1 if value else 0)

    @staticmethod
    def unpack(up):
        data = up.consume(1)
        return None if data is None else bool(data[0])


class list_type:
    _cache = dict()

    # todo: __getitem__ would be nice
    # todo: maybe even <T> syntax
    @staticmethod
    def of(elem_type):
        if elem_type not in list_type._cache:

            class list_type_inst:
                type_info = Pack(type_id.list_type.value, elem_type.type_info)

                @staticmethod
                def validate(value):
                    assert isinstance(value, list)
                    for elem in value:
                        elem_type.validate(elem)

                @staticmethod
                def pack(value):
                    list_type_inst.validate(value)
                    p = Packer()
                    p.pack_value(len(value), T=uint32_type)
                    for elem in value:
                        p.pack_value(elem, T=elem_type)
                    return p.data

                @staticmethod
                def unpack(up):
                    value = list()
                    n = up.unpack_value(T=uint32_type)
                    if n is None:
                        return None
                    for _ in range(n):
                        elem = up.unpack_value(T=elem_type)
                        if elem is None:
                            return None
                        value.append(elem)
                    return value

            list_type._cache[elem_type] = list_type_inst

        return list_type._cache[elem_type]


class string_type:
    type_info = Pack(type_id.string_type.value)

    @staticmethod
    def validate(value):
        assert isinstance(value, str)

    @staticmethod
    def pack(value):
        string_type.validate(value)
        p = Packer()
        p.pack_value(len(value), T=uint32_type)
        for ch in value:
            p.pack_value(ord(ch), T=uint8_type)
        return p.data

    @staticmethod
    def unpack(up):
        # todo: use StringIO for efficiency
        value = ""
        n = up.unpack_value(T=uint32_type)
        if n is None:
            return None
        for _ in range(n):
            ch = up.unpack_value(T=uint8_type)
            if ch is None:
                return None
            value += chr(ch)
        return value


def pack(value, T=None):
    T = T or resolve_type(value)
    return Packer().pack(value, T=T).data


def unpack_one(T, data):
    return Unpacker(data).unpack(T)


def unpack(Ts, data):
    if not isinstance(Ts, tuple):
        Ts = (Ts,)
    up = Unpacker(data)
    unpacked = tuple(up.unpack(T) for T in Ts)
    if any(value is None for value in unpacked):
        return None
    return unpacked


def test(value, T=None):
    T = T or resolve_type(value)
    print(f"input: {value!r}")

    packed = pack(value, T=T)
    print("packed:")
    packed.dump()

    unpacked = unpack_one(T, packed)
    if unpacked is None:
        print("failed to unpack")
    else:
        print(f"unpacked: {value!r}")


def main():
    test(5, T=uint32_type)
    print()
    test([1, 2, 3, 4, 5], T=list_type.of(float_type))
    print()
    test([-1, 1, -2, 2, -3, 3, -4, 4], T=list_type.of(int8_type))
    print()
    test("hello world")


if __name__ == "__main__":
    main()
